File: todo/frontend/src/services/todoService.ts
import axios from 'axios'; // 一个用于发送 HTTP 请求的库，常用于与 REST API 通信。
import type { TodoInfo, TodoCreate } from '../types/todo';

const API_URL = 'http://localhost:8001/api'; // fastapi 后端地址

export const todoService = {
    // async 说明是一个异步函数，会返回Promise，Promise<TodoInfo[]> 表示将来会返回一个包含TodoInfo Objects的list
    async getAll(): Promise<TodoInfo[]> {
        const response = await axios.get<TodoInfo[]>(`${API_URL}/todos`);
        return response.data
    },
    async getById(id: string): Promise<TodoInfo> {
        const response = await axios.get<TodoInfo>(`${API_URL}/todos/${id}`);
        return response.data
    },
    async create(todo: TodoCreate): Promise<TodoInfo> {
        // axios.post<ResponseType>(url, data, config?)
        // axios.post<你想要返回的类型>(url, 要发送的数据, 可选配置)
        const response = await axios.post<TodoInfo>(`${API_URL}/todos`, todo);
        return response.data
    },
    async update(id: string, todo: TodoCreate): Promise<TodoInfo> {
        const response = await axios.put<TodoInfo>(`${API_URL}/todos/${id}`, todo);
        return response.data
    },
    async delete(id: string): Promise<void> {
        await axios.delete(`${API_URL}/todos/${id}`); // 不需要 <TodoInfo> 泛型，因为不关心返回值
    }

}

File: todo/backend/app/services/todo_service.py
from typing import List, Optional
from uuid import UUID
from fastapi import HTTPException
from sqlalchemy.orm import Session
from app.repositories.todo_repository import TodoRepository
from app.schemas.todo import TodoCreate, TodoUpdate, TodoResponse

class TodoService:
    def __init__(self, db: Session):
        self.repository = TodoRepository(db)
    
    def get_all_todos(self) -> List[TodoResponse]:
        todos = self.repository.get_all()
        return [TodoResponse.from_orm(todo) for todo in todos]
    
    def get_todo_by_id(self, todo_id: UUID) -> TodoResponse:
        todo = self.repository.get_by_id(todo_id)
        if not todo:
            raise HTTPException(status_code=404, detail="Todo not found")
        return TodoResponse.from_orm(todo)
    
    def create_todo(self, todo_data: TodoCreate) -> TodoResponse:
        todo = self.repository.create(todo_data)
        return TodoResponse.from_orm(todo)
    
    def update_todo(self, todo_id: UUID, todo_data: TodoUpdate) -> TodoResponse:
        todo = self.repository.update(todo_id, todo_data)
        if not todo:
            raise HTTPException(status_code=404, detail="Todo not found")
        return TodoResponse.from_orm(todo)
    
    def delete_todo(self, todo_id: UUID) -> None:
        if not self.repository.delete(todo_id):
            raise HTTPException(status_code=404, detail="Todo not found")